commcare_connect/form_receiver/processor.py

```diff
@@ -248,67 +254,78 @@ def process_deliver_unit(user, xform: XForm, app: CommCareApp, opportunity: Oppo
+            entity_name=entity_name,
+            visit_date=xform.metadata.timeStart,
+            xform_id=xform.id,
+            app_build_id=xform.build_id,
+            app_build_version=xform.metadata.app_build_version,
+            form_json=xform.raw_form,
+            location=xform.metadata.location,
         )
-        user_visit.completed_work = completed_work
+        completed_work_needs_save = False
+        today = datetime.date.today()
+        paymentunit_startdate = payment_unit.start_date if payment_unit else None
+        if opportunity.start_date > today or (paymentunit_startdate and paymentunit_startdate > today):
+            completed_work = None
+            user_visit.status = VisitValidationStatus.trial
+        else:
+            completed_work, _ = CompletedWork.objects.get_or_create(
+                opportunity_access=access,
+                entity_id=entity_id,
+                payment_unit=payment_unit,
+                defaults={"entity_name": entity_name},
+            )
+            user_visit.completed_work = completed_work
+            if (
+                counts["daily"] >= payment_unit.max_daily
+                or counts["total"] >= claim_limit.max_visits
+                or (today > claim.end_date or (claim_limit.end_date and today > claim_limit.end_date))
+            ):
+                user_visit.status = VisitValidationStatus.over_limit
+                if not completed_work.status == CompletedWorkStatus.over_limit:
+                    completed_work.status = CompletedWorkStatus.over_limit
+                    completed_work_needs_save = True
+            elif counts["entity"] > 0:
+                user_visit.status = VisitValidationStatus.duplicate
+        flags = clean_form_submission(access, user_visit, xform)
+        if access.suspended:
+            flags.append(["user_suspended", "This user is suspended from the opportunity."])
+            user_visit.status = VisitValidationStatus.rejected
+        if flags:
+            user_visit.flagged = True
+            user_visit.flag_reason = {"flags": flags}
+
         if (
-            counts["daily"] >= payment_unit.max_daily
-            or counts["total"] >= claim_limit.max_visits
-            or (today > claim.end_date or (claim_limit.end_date and today > claim_limit.end_date))
+            opportunity.auto_approve_visits
+            and user_visit.status == VisitValidationStatus.pending
+            and not user_visit.flagged
         ):
-            user_visit.status = VisitValidationStatus.over_limit
-            if not completed_work.status == CompletedWorkStatus.over_limit:
-                completed_work.status = CompletedWorkStatus.over_limit
+            user_visit.status = VisitValidationStatus.approved
+            user_visit.review_status = VisitReviewStatus.agree
+
+        try:
+            user_visit.save()
+        except IntegrityError as e:
+            if "unique_xform_entity_deliver_unit" in str(e):
```

Comment:
Is there no other way to catch this? It is fine if not but kind of kludgy

Reply:
nit: I think we can remove this error-handling code here and let the unique constraint error be raised and propagated to the sentry.

Reply:
offlined with Hemant about this, but I prefer this to raising the error because it isn't an unexpected excpetion but one we are properly handling so there is no need to send to sentry. Instead we can log if we we want a record, but the key here is that we can return a meaningful error to HQ (and honestly probably want to return a 200, since we dont want them to keep retrying).

Reply:
Cool, that sounds good to me. I am happy with adding a log to show that a form was reforwarded from hq and returning a 200 status to HQ.